File: Vanilla/sources/zomboid/zombie/characters/SurvivorDesc.java
package zombie.characters;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map.Entry;
import se.krka.kahlua.vm.KahluaTable;
import zombie.GameWindow;
import zombie.Lua.LuaManager;
import zombie.characters.WornItems.BodyLocations;
import zombie.characters.WornItems.WornItems;
import zombie.characters.professions.ProfessionFactory;
import zombie.characters.skills.PerkFactory;
import zombie.characters.traits.ObservationFactory;
import zombie.core.Color;
import zombie.core.Rand;
import zombie.core.skinnedmodel.population.OutfitRNG;
import zombie.core.skinnedmodel.visual.HumanVisual;
import zombie.core.skinnedmodel.visual.IHumanVisual;
import zombie.core.skinnedmodel.visual.ItemVisuals;
import zombie.core.textures.ColorInfo;
import zombie.inventory.InventoryItem;
import zombie.iso.IsoWorld;

public final class SurvivorDesc implements IHumanVisual {
   public final HumanVisual humanVisual = new HumanVisual(this);
   public final WornItems wornItems = new WornItems(BodyLocations.getGroup("Human"));
   SurvivorGroup group = new SurvivorGroup();
   private static int IDCount = 0;
   public static final ArrayList TrouserCommonColors = new ArrayList();
   public static final ArrayList HairCommonColors = new ArrayList();
   private final HashMap xpBoostMap = new HashMap();
   private KahluaTable metaTable;
   public String Profession = "";
   protected String forename = "None";
   protected int ID = 0;
   protected IsoGameCharacter Instance = null;
   private boolean bFemale = true;
   protected String surname = "None";
   private String InventoryScript = null;
   protected String torso = "Base_Torso";
   protected final HashMap MetCount = new HashMap();
   protected float bravery = 1.0F;
   protected float loner = 0.0F;
   protected float aggressiveness = 1.0F;
   protected float compassion = 1.0F;
   protected float temper = 0.0F;
   protected float friendliness = 0.0F;
   private float favourindoors = 0.0F;
   protected float loyalty = 0.0F;
   public final ArrayList extra = new ArrayList();
   private final ArrayList Observations = new ArrayList(0);
   private SurvivorFactory.SurvivorType type;
   public boolean bDead;

   public HumanVisual getHumanVisual() {
      return this.humanVisual;
   }

   public void getItemVisuals(ItemVisuals var1) {
      this.wornItems.getItemVisuals(var1);
   }

   public boolean isFemale() {
      return this.bFemale;
   }

   public boolean isZombie() {
      return false;
   }

   public boolean isSkeleton() {
      return false;
   }

   public WornItems getWornItems() {
      return this.wornItems;
   }

   public void setWornItem(String var1, InventoryItem var2) {
      this.wornItems.setItem(var1, var2);
   }

   public InventoryItem getWornItem(String var1) {
      return this.wornItems.getItem(var1);
   }

   public void dressInNamedOutfit(String var1) {
      ItemVisuals var2 = new ItemVisuals();
      this.getHumanVisual().dressInNamedOutfit(var1, var2);
      this.getWornItems().setFromItemVisuals(var2);
   }

   public SurvivorGroup getGroup() {
      return this.group;
   }

   public boolean isLeader() {
      return this.group.getLeader() == this;
   }

   public static int getIDCount() {
      return IDCount;
   }

   public void setProfessionSkills(ProfessionFactory.Profession var1) {
      this.getXPBoostMap().clear();
      this.getXPBoostMap().putAll(var1.XPBoostMap);
   }

   public HashMap getXPBoostMap() {
      return this.xpBoostMap;
   }

   public KahluaTable getMeta() {
      if (this.metaTable == null) {
         this.metaTable = (KahluaTable)LuaManager.caller.pcall(LuaManager.thread, LuaManager.env.rawget("createMetaSurvivor"), (Object)this)[1];
      }

      return this.metaTable;
   }

   public int getCalculatedToughness() {
      this.metaTable = this.getMeta();
      KahluaTable var1 = (KahluaTable)LuaManager.env.rawget("MetaSurvivor");
      Double var2 = (Double)LuaManager.caller.pcall(LuaManager.thread, var1.rawget("getCalculatedToughness"), (Object)this.metaTable)[1];
      return var2.intValue();
   }

   public static void setIDCount(int var0) {
      IDCount = var0;
   }

   public boolean isDead() {
      return this.bDead;
   }

   public SurvivorDesc() {
      this.type = SurvivorFactory.SurvivorType.Neutral;
      this.ID = IDCount++;
      IsoWorld.instance.SurvivorDescriptors.put(this.ID, this);
      this.doStats();
   }

   public SurvivorDesc(boolean var1) {
      this.type = SurvivorFactory.SurvivorType.Neutral;
      this.ID = IDCount++;
      this.doStats();
   }

   public SurvivorDesc(SurvivorDesc var1) {
      this.type = SurvivorFactory.SurvivorType.Neutral;
      this.aggressiveness = var1.aggressiveness;
      this.bDead = var1.bDead;
      this.bFemale = var1.bFemale;
      this.bravery = var1.bravery;
      this.compassion = var1.compassion;
      this.extra.addAll(var1.extra);
      this.favourindoors = var1.favourindoors;
      this.forename = var1.forename;
      this.friendliness = var1.friendliness;
      this.InventoryScript = var1.InventoryScript;
      this.loner = var1.loner;
      this.loyalty = var1.loyalty;
      this.Profession = var1.Profession;
      this.surname = var1.surname;
      this.temper = var1.temper;
      this.torso = var1.torso;
      this.type = var1.type;
   }

   public void meet(SurvivorDesc var1) {
      if (this.MetCount.containsKey(var1.ID)) {
         this.MetCount.put(var1.ID, (Integer)this.MetCount.get(var1.ID) + 1);
      } else {
         this.MetCount.put(var1.ID, 1);
      }

      if (var1.MetCount.containsKey(this.ID)) {
         var1.MetCount.put(this.ID, (Integer)var1.MetCount.get(this.ID) + 1);
      } else {
         var1.MetCount.put(this.ID, 1);
      }

   }

   public boolean hasObservation(String var1) {
      for(int var2 = 0; var2 < this.Observations.size(); ++var2) {
         if (var1.equals(((ObservationFactory.Observation)this.Observations.get(var2)).getTraitID())) {
            return true;
         }
      }

      return false;
   }

   private void savePerk(ByteBuffer var1, PerkFactory.Perk var2) throws IOException {
      GameWindow.WriteStringUTF(var1, var2 == null ? "" : var2.getId());
   }

   private PerkFactory.Perk loadPerk(ByteBuffer var1, int var2) throws IOException {
      PerkFactory.Perk var4;
      if (var2 >= 152) {
         String var5 = GameWindow.ReadStringUTF(var1);
         var4 = PerkFactory.Perks.FromString(var5);
         return var4 == PerkFactory.Perks.MAX ? null : var4;
      } else {
         int var3 = var1.getInt();
         if (var3 >= 0 && var3 < PerkFactory.Perks.MAX.index()) {
            var4 = PerkFactory.Perks.fromIndex(var3);
            return var4 == PerkFactory.Perks.MAX ? null : var4;
         } else {
            return null;
         }
      }
   }

   public void load(ByteBuffer var1, int var2, IsoGameCharacter var3) throws IOException {
      this.ID = var1.getInt();
      IsoWorld.instance.SurvivorDescriptors.put(this.ID, this);
      this.forename = GameWindow.ReadString(var1);
      this.surname = GameWindow.ReadString(var1);
      this.torso = GameWindow.ReadString(var1);
      this.bFemale = var1.getInt() == 1;
      this.Profession = GameWindow.ReadString(var1);
      this.doStats();
      if (IDCount < this.ID) {
         IDCount = this.ID;
      }

      this.extra.clear();
      int var4;
      int var5;
      if (var1.getInt() == 1) {
         var4 = var1.getInt();

         for(var5 = 0; var5 < var4; ++var5) {
            String var6 = GameWindow.ReadString(var1);
            this.extra.add(var6);
         }
      }

      var4 = var1.getInt();

      for(var5 = 0; var5 < var4; ++var5) {
         PerkFactory.Perk var8 = this.loadPerk(var1, var2);
         int var7 = var1.getInt();
         if (var8 != null) {
            this.getXPBoostMap().put(var8, var7);
         }
      }

      this.Instance = var3;
   }

   public void save(ByteBuffer var1) throws IOException {
      var1.putInt(this.ID);
      GameWindow.WriteString(var1, this.forename);
      GameWindow.WriteString(var1, this.surname);
      GameWindow.WriteString(var1, this.torso);
      var1.putInt(this.bFemale ? 1 : 0);
      GameWindow.WriteString(var1, this.Profession);
      if (!this.extra.isEmpty()) {
         var1.putInt(1);
         var1.putInt(this.extra.size());

         for(int var2 = 0; var2 < this.extra.size(); ++var2) {
            String var3 = (String)this.extra.get(var2);
            GameWindow.WriteString(var1, var3);
         }
      } else {
         var1.putInt(0);
      }

      var1.putInt(this.getXPBoostMap().size());
      Iterator var4 = this.getXPBoostMap().entrySet().iterator();

      while(var4.hasNext()) {
         Entry var5 = (Entry)var4.next();
         this.savePerk(var1, (PerkFactory.Perk)var5.getKey());
         var1.putInt((Integer)var5.getValue());
      }

   }

   public void loadCompact(ByteBuffer var1) {
      this.ID = -1;
      this.torso = GameWindow.ReadString(var1);
      this.bFemale = var1.get() == 1;
      this.extra.clear();
      if (var1.get() == 1) {
         byte var2 = var1.get();

         for(int var3 = 0; var3 < var2; ++var3) {
            String var4 = GameWindow.ReadString(var1);
            this.extra.add(var4);
         }
      }

   }

   public void saveCompact(ByteBuffer var1) throws UnsupportedEncodingException {
      GameWindow.WriteString(var1, this.torso);
      var1.put((byte)(this.bFemale ? 1 : 0));
      if (!this.extra.isEmpty()) {
         var1.put((byte)1);
         var1.put((byte)this.extra.size());
         Iterator var2 = this.extra.iterator();

         while(var2.hasNext()) {
            String var3 = (String)var2.next();
            GameWindow.WriteString(var1, var3);
         }
      } else {
         var1.put((byte)0);
      }

   }

   public void addObservation(String var1) {
      ObservationFactory.Observation var2 = ObservationFactory.getObservation(var1);
      if (var2 != null) {
         this.Observations.add(var2);
      }
   }

   private void doStats() {
      this.bravery = Rand.Next(2) == 0 ? 10.0F : 0.0F;
      this.aggressiveness = Rand.Next(2) == 0 ? 10.0F : 0.0F;
      this.compassion = 10.0F - this.aggressiveness;
      this.loner = Rand.Next(2) == 0 ? 10.0F : 0.0F;
      this.temper = Rand.Next(2) == 0 ? 10.0F : 0.0F;
      this.friendliness = 10.0F - this.loner;
      this.favourindoors = Rand.Next(2) == 0 ? 10.0F : 0.0F;
      this.loyalty = Rand.Next(2) == 0 ? 10.0F : 0.0F;
   }

   public int getMetCount(SurvivorDesc var1) {
      return this.MetCount.containsKey(var1.ID) ? (Integer)this.MetCount.get(var1.ID) : 0;
   }

   public String getForename() {
      return this.forename;
   }

   public void setForename(String var1) {
      this.forename = var1;
   }

   public int getID() {
      return this.ID;
   }

   public void setID(int var1) {
      this.ID = var1;
   }

   public IsoGameCharacter getInstance() {
      return this.Instance;
   }

   public void setInstance(IsoGameCharacter var1) {
      this.Instance = var1;
   }

   public String getSurname() {
      return this.surname;
   }

   public void setSurname(String var1) {
      this.surname = var1;
   }

   public String getInventoryScript() {
      return this.InventoryScript;
   }

   public void setInventoryScript(String var1) {
      this.InventoryScript = var1;
   }

   public String getTorso() {
      return this.torso;
   }

   public void setTorso(String var1) {
      this.torso = var1;
   }

   public HashMap getMetCount() {
      return this.MetCount;
   }

   public float getBravery() {
      return this.bravery;
   }

   public void setBravery(float var1) {
      this.bravery = var1;
   }

   public float getLoner() {
      return this.loner;
   }

   public void setLoner(float var1) {
      this.loner = var1;
   }

   public float getAggressiveness() {
      return this.aggressiveness;
   }

   public void setAggressiveness(float var1) {
      this.aggressiveness = var1;
   }

   public float getCompassion() {
      return this.compassion;
   }

   public void setCompassion(float var1) {
      this.compassion = var1;
   }

   public float getTemper() {
      return this.temper;
   }

   public void setTemper(float var1) {
      this.temper = var1;
   }

   public float getFriendliness() {
      return this.friendliness;
   }

   public void setFriendliness(float var1) {
      this.friendliness = var1;
   }

   public float getFavourindoors() {
      return this.favourindoors;
   }

   public void setFavourindoors(float var1) {
      this.favourindoors = var1;
   }

   public float getLoyalty() {
      return this.loyalty;
   }

   public void setLoyalty(float var1) {
      this.loyalty = var1;
   }

   public String getProfession() {
      return this.Profession;
   }

   public void setProfession(String var1) {
      this.Profession = var1;
   }

   public boolean isAggressive() {
      Iterator var1 = this.Observations.iterator();

      ObservationFactory.Observation var2;
      do {
         if (!var1.hasNext()) {
            return false;
         }

         var2 = (ObservationFactory.Observation)var1.next();
      } while(!"Aggressive".equals(var2.getTraitID()));

      return true;
   }

   public ArrayList getObservations() {
      return this.Observations;
   }

   public boolean isFriendly() {
      Iterator var1 = this.Observations.iterator();

      ObservationFactory.Observation var2;
      do {
         if (!var1.hasNext()) {
            return false;
         }

         var2 = (ObservationFactory.Observation)var1.next();
      } while(!"Friendly".equals(var2.getTraitID()));

      return true;
   }

   public SurvivorFactory.SurvivorType getType() {
      return this.type;
   }

   public void setType(SurvivorFactory.SurvivorType var1) {
      this.type = var1;
   }

   public void setFemale(boolean var1) {
      this.bFemale = var1;
   }

   public ArrayList getExtras() {
      return this.extra;
   }

   public ArrayList getCommonHairColor() {
      return HairCommonColors;
   }

   public static void addTrouserColor(ColorInfo var0) {
      TrouserCommonColors.add(var0.toColor());
   }

   public static void addHairColor(ColorInfo var0) {
      HairCommonColors.add(var0.toImmutableColor());
   }

   public static Color getRandomSkinColor() {
      return OutfitRNG.Next(3) == 0 ? new Color(OutfitRNG.Next(0.5F, 0.6F), OutfitRNG.Next(0.3F, 0.4F), OutfitRNG.Next(0.15F, 0.23F)) : new Color(OutfitRNG.Next(0.9F, 1.0F), OutfitRNG.Next(0.75F, 0.88F), OutfitRNG.Next(0.45F, 0.58F));
   }
}

File: Vanilla/sources/zdoc/media/lua/shared/Library/SurvivorDesc.lua
---@class SurvivorDesc : zombie.characters.SurvivorDesc
---@field public humanVisual HumanVisual
---@field public wornItems WornItems
---@field group SurvivorGroup
---@field private IDCount int
---@field public TrouserCommonColors ArrayList|Color
---@field public HairCommonColors ArrayList|Color
---@field public xpBoostMap HashMap|PerkFactory.Perks|Integer
---@field private metaTable KahluaTable
---@field public Profession String
---@field protected forename String
---@field protected ID int
---@field protected Instance IsoGameCharacter
---@field private bFemale boolean
---@field protected surname String
---@field private InventoryScript String
---@field protected torso String
---@field protected MetCount HashMap|Unknown|Unknown
---@field protected bravery float
---@field protected loner float
---@field protected aggressiveness float
---@field protected compassion float
---@field protected temper float
---@field protected friendliness float
---@field private favourindoors float
---@field protected loyalty float
---@field public extra ArrayList|String
---@field public Observations ArrayList|ObservationFactory.Observation
---@field private type SurvivorFactory.SurvivorType
---@field public bDead boolean
SurvivorDesc = {}

---@public
---@param forename String @the forename to set
---@return void
function SurvivorDesc:setForename(forename) end

---@public
---@return HashMap|Integer|Integer
---@overload fun(descriptor:SurvivorDesc)
function SurvivorDesc:getMetCount() end

---@public
---@param descriptor SurvivorDesc
---@return int
function SurvivorDesc:getMetCount(descriptor) end

---@public
---@return IsoGameCharacter @the Instance
function SurvivorDesc:getInstance() end

---@public
---@param arg0 String
---@return void
function SurvivorDesc:dressInNamedOutfit(arg0) end

---throws java.io.IOException
---@public
---@param input ByteBuffer
---@return void
function SurvivorDesc:loadCompact(input) end

---@private
---@param arg0 ByteBuffer
---@param arg1 PerkFactory.Perk
---@return void
function SurvivorDesc:savePerk(arg0, arg1) end

---@public
---@param desc SurvivorDesc
---@return void
function SurvivorDesc:meet(desc) end

---@public
---@param ID int @the ID to set
---@return void
function SurvivorDesc:setID(ID) end

---@public
---@param obv String
---@return void
function SurvivorDesc:addObservation(obv) end

---@public
---@param loner float @the loner to set
---@return void
function SurvivorDesc:setLoner(loner) end

---@public
---@return float @the bravery
function SurvivorDesc:getBravery() end

---@public
---@return String @the torso
function SurvivorDesc:getTorso() end

---@public
---@return int @the IDCount
function SurvivorDesc:getIDCount() end

---@public
---@return ArrayList|String
function SurvivorDesc:getExtras() end

---@public
---@return boolean
function SurvivorDesc:isLeader() end

---throws java.io.IOException
---@public
---@param input ByteBuffer
---@param WorldVersion int
---@param chr IsoGameCharacter
---@return void
function SurvivorDesc:load(input, WorldVersion, chr) end

---@public
---@return float @the friendliness
function SurvivorDesc:getFriendliness() end

---@public
---@param surname String @the surname to set
---@return void
function SurvivorDesc:setSurname(surname) end

---@public
---@param temper float @the temper to set
---@return void
function SurvivorDesc:setTemper(temper) end

---@public
---@return SurvivorFactory.SurvivorType
function SurvivorDesc:getType() end

---@public
---@return HashMap|PerkFactory.Perks|Integer
function SurvivorDesc:getXPBoostMap() end

---@public
---@return boolean
function SurvivorDesc:isSkeleton() end

---@public
---@param Instance IsoGameCharacter @the Instance to set
---@return void
function SurvivorDesc:setInstance(Instance) end

---@public
---@param profession ProfessionFactory.Profession
---@return void
function SurvivorDesc:setProfessionSkills(profession) end

---@public
---@param arg0 String
---@return InventoryItem
function SurvivorDesc:getWornItem(arg0) end

---@public
---@return float @the compassion
function SurvivorDesc:getCompassion() end

---@public
---@param compassion float @the compassion to set
---@return void
function SurvivorDesc:setCompassion(compassion) end

---@public
---@return String @the InventoryScript
function SurvivorDesc:getInventoryScript() end

---@public
---@return Color
function SurvivorDesc:getRandomSkinColor() end

---@public
---@return float @the aggressiveness
function SurvivorDesc:getAggressiveness() end

---@private
---@param arg0 ByteBuffer
---@param arg1 int
---@return PerkFactory.Perk
function SurvivorDesc:loadPerk(arg0, arg1) end

---@public
---@return float @the temper
function SurvivorDesc:getTemper() end

---@public
---@return float @the favourindoors
function SurvivorDesc:getFavourindoors() end

---@public
---@return boolean
function SurvivorDesc:isFriendly() end

---@public
---@param aIDCount int @the IDCount to set
---@return void
function SurvivorDesc:setIDCount(aIDCount) end

---throws java.io.IOException
---@public
---@param output ByteBuffer
---@return void
function SurvivorDesc:save(output) end

---@public
---@return HumanVisual
function SurvivorDesc:getHumanVisual() end

---@public
---@return ArrayList|ObservationFactory.Observation
function SurvivorDesc:getObservations() end

---@public
---@return SurvivorGroup @the Group
function SurvivorDesc:getGroup() end

---@public
---@param color ColorInfo
---@return void
function SurvivorDesc:addHairColor(color) end

---@public
---@param color ColorInfo
---@return void
function SurvivorDesc:addTrouserColor(color) end

---@public
---@return int @the ID
function SurvivorDesc:getID() end

---@public
---@return String @the Profession
function SurvivorDesc:getProfession() end

---@public
---@return float @the loyalty
function SurvivorDesc:getLoyalty() end

---@public
---@return String @the forename
function SurvivorDesc:getForename() end

---@public
---@param loyalty float @the loyalty to set
---@return void
function SurvivorDesc:setLoyalty(loyalty) end

---@public
---@return String @the surname
function SurvivorDesc:getSurname() end

---@private
---@return void
function SurvivorDesc:doStats() end

---@public
---@return ArrayList|Color
function SurvivorDesc:getCommonHairColor() end

---@public
---@return boolean
function SurvivorDesc:isAggressive() end

---@public
---@return int
function SurvivorDesc:getCalculatedToughness() end

---@public
---@return boolean
function SurvivorDesc:isZombie() end

---@public
---@param arg0 ItemVisuals
---@return void
function SurvivorDesc:getItemVisuals(arg0) end

---@public
---@param aggressiveness float @the aggressiveness to set
---@return void
function SurvivorDesc:setAggressiveness(aggressiveness) end

---@public
---@param Profession String @the Profession to set
---@return void
function SurvivorDesc:setProfession(Profession) end

---@public
---@param favourindoors float @the favourindoors to set
---@return void
function SurvivorDesc:setFavourindoors(favourindoors) end

---@public
---@param friendliness float @the friendliness to set
---@return void
function SurvivorDesc:setFriendliness(friendliness) end

---@public
---@return WornItems
function SurvivorDesc:getWornItems() end

---@public
---@return KahluaTable
function SurvivorDesc:getMeta() end

---@public
---@param bFemale boolean
---@return void
function SurvivorDesc:setFemale(bFemale) end

---@public
---@return boolean
function SurvivorDesc:isFemale() end

---@public
---@param o String
---@return boolean
function SurvivorDesc:hasObservation(o) end

---@public
---@param bravery float @the bravery to set
---@return void
function SurvivorDesc:setBravery(bravery) end

---@public
---@param arg0 String
---@param arg1 InventoryItem
---@return void
function SurvivorDesc:setWornItem(arg0, arg1) end

---@public
---@param InventoryScript String @the InventoryScript to set
---@return void
function SurvivorDesc:setInventoryScript(InventoryScript) end

---@public
---@param type SurvivorFactory.SurvivorType
---@return void
function SurvivorDesc:setType(type) end

---throws java.io.IOException
---@public
---@param output ByteBuffer
---@return void
function SurvivorDesc:saveCompact(output) end

---@public
---@param torso String @the torso to set
---@return void
function SurvivorDesc:setTorso(torso) end

---@public
---@return boolean
function SurvivorDesc:isDead() end

---@public
---@return float @the loner
function SurvivorDesc:getLoner() end
